Implicit tiling spec

ptrgags opened the pull request:
[Direct link to `3DTILES_implicit_tiling` spec](https://github.com/CesiumGS/3d-tiles/tree/implicit-tiling-spec/extensions/3DTILES_implicit_tiling)

Opening this early because I wanted to refer to some of the availability buffer formulas in another issue.

This is the specification to describe the implicit tiling schema we merged into `3d-tiles-next` last week.

I'll update this description once I have more of the text fleshed out.

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/schema/subtree.schema.json:
```diff
@@ -19,17 +19,17 @@
     },
     "tileAvailability": {
-      "description": "The availability of tiles in the subtree. If not all available or unavailable, the size of the buffer will be (N ^ (subtreeDepth + 1) - 1) / (N - 1) bits where N is 4 for tiling scheme `QUADTREE` and 8 for `OCTREE`.",
+      "description": "The availability of tiles in the subtree. If not all available or unavailable, the size of the buffer will be (N ^ (subtreeLevels + 1) - 1) / (N - 1) bits where N is 4 for tiling scheme `QUADTREE` and 8 for `OCTREE`.",
```
@ErixenCruz I think this formula has an off-by-one error due to how `subtreeLevels` is defined (note that the terminology changed recently) I think it should just be `subtreeLevels` without the +1

Here's my thinking:

| `subtreeLevels` | bits |
|----|---|
| 1 | 1|
| 2 | 1 + 4 = 5 |
| 3 | 1 + 4 + 16 = 21 |
| L | sum for k in `[0, L -1]` of N^k = (N^L - 1) / (N - 1) |

Could you double check this? Then I'll update the schema and tables in the spec.

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/schema/subtree.schema.json:
```diff
@@ -19,17 +19,17 @@
     },
     "contentAvailability": {
-      "description": "The availability of the content of the tiles of the subtree. If not all available or unavailable, the size of the buffer will be N ^(subtreeDepth) bits where N is 4 for tiling scheme `QUADTREE` and 8 for `OCTREE`.",
+      "description": "The availability of the content of the tiles of the subtree. If not all available or unavailable, the size of the buffer will be N ^(subtreeLevels) bits where N is 4 for tiling scheme `QUADTREE` and 8 for `OCTREE`.",
```
Once the other comment is addressed, I'll update this to match `tileAvailability` since these two buffers must match.

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -19,19 +19,242 @@ Written against the 3D Tiles 1.0 specification.
+
+| Quantity | Formula | Description |
+| -------- | ------- | ----------- |
+| `lengthBits` | `N^subtreeLevels - 1` | Length of buffer is determined by subtree levels 
```
Incorrect, this should be `(N^subtreeLevels - 1) / (N - 1)`

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -19,19 +19,242 @@ Written against the 3D Tiles 1.0 specification.
+| `parent.index` | `floor((child.index - 1) / N)` | Index of the parent in the bit vector | 
+| `parent.indexOf(child)` | `(child.index - 1) % N` | Index of the child within the parent's `N` children |
+| `parent.children[k].index` | `N * index + k + 1` | Find the bit of the `k-th` child of a node |
+| `index` | `N^level - 1 + mortonIndex` | Find the index of a node from `(level, mortonIndex)`
```
I need to double check these formulae involving N^level.

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/example/backup.json:
```diff
@@ -0,0 +1,68 @@
+{
```
Whoops I didn't mean to commit this.

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/example/everything.json:
```diff
@@ -94,8 +94,8 @@
         "geometricError": 5000,
-        "subtreeDepth": 7,
-        "maximumDepth": 21,
+        "subtreeLevels": 7,
+        "maximumLevel": 21,
```
@ErixenCruz these are the two renaming changes of depth -> level for clarity.

`subtreeLevels` is how many distinct levels in the tree there are. So for a quadtree with 2 levels (1 + 4 nodes), `subtreeLevels = 2`.

`maximumLevel` is the 0-indexed level of the last valid level. I know this sounds weird, as "depth" is _technically_ more accurate, but  Cesium 3D Tiles starts at level = 0 so we need to follow this convention.
![image](https://user-images.githubusercontent.com/8422414/101515353-430aa180-394c-11eb-9472-3d7411d5cb3f.png)

ptrgags reviewed (commented):
Added some notes about problems with formulas

ptrgags commented:
@sanjeetsuhag I just pushed the commits @ErixenCruz and I have been working on. There's still a couple sections and diagrams in progress, but you can review what we have so far

sanjeetsuhag commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,522 @@
+
+>**Template URIs** are URI patterns used to refer to specific tiles by their tile coordinates.
+
+Template URIs are configured in the tileset.json. They may be any URI pattern, but must include the variables `${level}`, `${x}`, `${x}`, and for octrees also `${z}`. When referring to a specific tile, the tile's coordinates are substituted in for these variables.
```
```suggestion
Template URIs are configured in the tileset.json. They may be any URI pattern, but must include the variables `${level}`, `${x}`, `${y}`, and for octrees also `${z}`. When referring to a specific tile, the tile's coordinates are substituted in for these variables.
```

sanjeetsuhag commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,522 @@
+
+This extension adds the `mimeType` property to identify the type of content. This is more reliable than a file extension for determining the file type.
+
+To use glTF models (`.gltf` files) or binary glTF models (`.glb` files) as content, the `3DTILES_content_gltf` extension must be listed as a required extension in the tileset.
```
This may not belong in this spec.

sanjeetsuhag commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,522 @@
+1. It provides an efficient method for checking what files are present
+2. Including ths information prevents extraneous HTTP requests that would result in 404 errors.
+
+Availablity takes the form of a bitstream with one bit per node in consideration. A 1 indicates that a tile/content/subtree is available at this node. Meanwhile, a 0 indicates that no data is available.
```
```suggestion
Availablity takes the form of a bitstream with one bit per node in consideration. A 1 indicates that a tile/content/subtree is available at this node. Meanwhile, a 0 indicates that a tile/content/subtree is not available at this node.
```

I would be explicit about this here, even though it risks being too verbose. "No data" sounded vague to me.

sanjeetsuhag commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,522 @@
+
+Availablity takes the form of a bitstream with one bit per node in consideration. A 1 indicates that a tile/content/subtree is available at this node. Meanwhile, a 0 indicates that no data is available.
+
+When every node is available or every node is unavailable, all the bits of the bitstream will be identical. That is, either all 1s or all 0s. Instead of storing a full bit stream, the `constant` property can be used instead. For example, `constant: 0` indicates that all bits are 0 and no bitstream must be stored.
```
```suggestion
For cases where all node are available or all nodes are unavailable, the `constant` property should be used instead of storing a full bitstream. For example, `constant: 0` indicates that all bits are 0 and no bitstream must be stored.
```

sanjeetsuhag commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,522 @@
+
+The purpose of content availability is to check if a file exists before making a network request. If content is marked as unavailable, the network request for that file must be skipped.
+
+A content availablity bit must only be set if the corresponding tile availability bit is set. Otherwise, it would be possible to specify content files that are not reachable by the tiles of the tileset. The content availability bitstream can be validiated by checking that the following equation holds true:
```
```suggestion
A content availablity bit must only be set if the corresponding tile availability bit is set. Otherwise, it would be possible to specify content files that are not reachable by the tiles of the tileset. The content availability bitstream can be validated by checking that the following equation holds true:
```

sanjeetsuhag commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,522 @@
+| `boundingVolume` | a bounding volume (either a `box` or `region`) describing the root tile |
+| `refine` | Either `ADD` or `REPLACE` as in the Cesium 3D Tiles 1.0 Specification. |
+| `geometricError` | Geometric error of the root tile as described in the Cesium 3D Tiles 1.0 Specification. |
+| `maximumLevel` | Maxium level of the entire tree |
```
```suggestion
| `maximumLevel` | Maximum level of the entire tree |
```

sanjeetsuhag commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,522 @@
+| `refine` | Either `ADD` or `REPLACE` as in the Cesium 3D Tiles 1.0 Specification. |
+| `geometricError` | Geometric error of the root tile as described in the Cesium 3D Tiles 1.0 Specification. |
+| `maximumLevel` | Maxium level of the entire tree |
+| `subtreeLevels` | How many distince levels in each subtree. |
```
```suggestion
| `subtreeLevels` | How many distance levels in each subtree. |
```

sanjeetsuhag commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,522 @@
+* **octree** - A 3D tiling scheme that divides each cuboid into 8 smaller cuboids.
+* **quadtree** - A 2D tiling scheme that divides each rectangle into 4 smaller rectangles.
+* **root tile** - The topmost tile in a tileset tree.
+* **subtree** - A fixed-depth section of the tileset tree used to break large tilesets into managable pieces.
```
```suggestion
* **subtree** - A fixed-depth section of the tileset tree used to break large tilesets into manageable pieces.
```

sanjeetsuhag commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,522 @@
+| `y` | The 0-indexed number of tiles from the side of the root bounding box in the `+y` direction. This is measured within a single level.|
+| `z` (only for octrees) | The 0-indexed number of tiles from the side of the root bounding box in the `+z` direction. This is measured within a single level. |
+
+![Box coordinates](figures/box-coordinates.jpg)
```
Template URIs are being used in this diagram, so we need to either use the tuple form, or move the template URI section up. Another solution could be to add a legend at the top of the diagram indicating how to interpret the tile labels.

sanjeetsuhag commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,522 @@
+![subtree anatomy](figures/subtree-anatomy.jpg)
+
+OUTLINE:
+- refering to nodes is confusing. review this!
```
```suggestion
- referring to nodes is confusing. review this!
```

sanjeetsuhag commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,522 @@
+
+**Tile availability** is a bitstream that determines which tiles exist within a subtree. There is one bit for each subtree node. A 1 indicates that a tile is available, while a 0 indicates that a tile is unavailable.
+
+![Tile Availability](figures/tile-availability.jpg)
```
This diagram may potentially mislead someone into believing that binary trees are supported. We should use quadtrees.

ptrgags replied:
Yeah, my only hesitation is that a quadtree diagram could get quite a bit wider (since we need to show the entire buffer). 

As a group we decided to hold off for now and find a better solution later.

ptrgags commented:
@sanjeetsuhag updated!

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+<!-- omit in toc -->
 ## Contributors
 
 * Sam Suhag, Cesium
 * Sean Lilley, Cesium
 * Peter Gagliardi, Cesium
 * Erixen Cruz, Cesium
+* Patrick Cozzi, Cesium
```
We should revisit the order of authors now that the text of this draft is written.

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+<!-- omit in toc -->
 # 3DTILES_implicit_tiling
 
 **Version 0.0.0**, December 1, 2020
```
What date should we put here? We technically finished the draft on the 11th, but we could also wait until the day we merge this PR. 

Just remember to update this on the landing page too.

ErixenCruz commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
 ## Overview
 
-This extension to 3D Tiles enables implicit tiling. 
+>**Implicit tiling** is a alternative method for describing a Cesium 3D Tileset that provides a more succinct representation for large tilesets. It encodes the hierarchical structure of the tileset in binary, as opposed to explicit tiling, which specifies the tile hierarchy in JSON. The Cesium 3D Tiles 1.0 specification only supports explicit tiling, as every tile is listed in the tileset JSON fi ...
```
```suggestion
>**Implicit tiling** is an alternative method for describing a Cesium 3D Tileset that provides a more succinct representation for large tilesets. It encodes the hierarchical structure of the tileset in binary, as opposed to explicit tiling, which specifies the tile hierarchy in JSON. The Cesium 3D Tiles 1.0 specification only supports explicit tiling, as every tile is listed in the tileset JSON file.
```

ErixenCruz commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+
+Implicit tiling also provides a method for accessing tiles by tile coordinates. This allows for abbreviated tree traversal algorithms.
+
+![Explicit vs Implicit Tiling](figures/implicit-vs-explicit.jpg)
```
Should clarify binary tree is for visualization only since we only support quad and oct.

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+* [WMTS](https://www.ogc.org/standards/wmts)
+* [TMS](https://wiki.osgeo.org/wiki/Tile_Map_Service_Specification)
+
+One new feature implicit tiling enables is procedurally-generated tilesets. Since implicit tiling encodes tile coordinates in URLs (such as `{level}/{x}/{y}/model.gltf`), consider the server that serves these files. Instead of serving static files, a server could extract the tile coordinates from the URL and generate tiles at runtime. This could be useful for making a large procedural terrain dat ...
```
This paragraph could use some rewording. Also we might want to be careful about talking about template URIs since they are not described until a later section.

sanjeetsuhag replied:
```suggestion
One new feature implicit tiling enables is procedurally-generated tilesets. Since implicit tiling encodes tile coordinates in URLs using [Template URIs](#template-uris), consider the server that serves these files. Instead of serving static files, a server could extract the tile coordinates from the URL and generate tiles at runtime. This could be useful for making a large procedural terrain dataset without requiring much disk space.
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+* [WMTS](https://www.ogc.org/standards/wmts)
+* [TMS](https://wiki.osgeo.org/wiki/Tile_Map_Service_Specification)
+
+One new feature implicit tiling enables is procedurally-generated tilesets. Since implicit tiling encodes tile coordinates in URLs (such as `{level}/{x}/{y}/model.gltf`), consider the server that serves these files. Instead of serving static files, a server could extract the tile coordinates from the URL and generate tiles at runtime. This could be useful for making a large procedural terrain dat ...
```
Also, probably not wise to mention gltf content, that goes beyond the scope of this spec.

ErixenCruz commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+
+## Tile Coordinates
+
+>**Tile coordinates** are a tuple of integers that uniquely identify a tile. Tile coordinates are either `(level, x, y)` for quadtrees, and `(level, x, y, z)` for octrees.
```
```suggestion
>**Tile coordinates** are a tuple of integers that uniquely identify a tile. Tile coordinates are either `(level, x, y)` for quadtrees or `(level, x, y, z)` for octrees.
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+| Coordinate | Description |
+| --- | --- |
+| `level` | The 0-indexed level number. The root tile is at level 0, the root tile's children are at level 1, and so on. |
+| `x` | The 0-indexed number of tiles from the side of the root bounding box in the `+x` direction. This is measured within a single level. |
+| `y` | The 0-indexed number of tiles from the side of the root bounding box in the `+y` direction. This is measured within a single level.|
+| `z` (only for octrees) | The 0-indexed number of tiles from the side of the root bounding box in the `+z` direction. This is measured within a single level. |
```
Reword these a little.  should be "number of tiles from the side of the root bounding box in order of increasing x coordinate"

However, if we can find a more concise way to express this, that would be even better.

sanjeetsuhag replied:
Instead of adding verbosity in the coordinate descriptions, consider adding some text immediately before or after this table noting that all coordinates are 0 indexed.

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+
+>**Template URIs** are URI patterns used to refer to specific tiles by their tile coordinates.
+
+Template URIs are configured in the tileset.json. They may be any URI pattern, but must include the variables `${level}`, `${x}`, `${y}`, and for octrees also `${z}`. When referring to a specific tile, the tile's coordinates are substituted in for these variables.
```
```suggestion
Template URIs are configured in the tileset.json. They may be any URI pattern, but must include the variables `${level}`, `${x}`, `${y}`. Template URIs for octrees must also include `${z}`. When referring to a specific tile, the tile's coordinates are substituted in for these variables.
```

ErixenCruz commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
````diff
@@ -1,37 +1,739 @@
+```
+## Glossary
+
+* **availability** - Data specifying which tiles are available within a single subtree. This helps prevent unnecessary network requests.
````
```suggestion
* **availability** - Data specifying which tiles/subtrees/contents are available within a single subtree. This helps prevent unnecessary network requests.
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+
+![Child Subtree Availability](figures/subtree-availability.jpg)
+
+Child subtree availability is used to determine whether further subtree files exist before making network requests. If a child subtree availability bit is 0, any network request for that subtree must be skipped.
```
I find this wording somewhat ambiguous. "further" isn't the right word here.

sanjeetsuhag replied:
An alternative phrasing could be:

```suggestion
Child subtree availability is used to determine whether files for child subtrees exist before making network requests. If a child subtree availability bit is 0, any network request for that subtree must be skipped.
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+### Buffers and Buffer Views
+
+OUTLINE:
+- Which spec to reference? 3DTILES_binary_buffers or core metadata?
```
@sanjeetsuhag @ErixenCruz we should discuss this today.

ptrgags replied:
This should reference 3DTILES_binary_buffers, though that [draft spec](https://github.com/CesiumGS/3d-tiles/tree/3DTILES_binary_buffers/extensions/3DTILES_binary_buffers) doesn't really explain what buffers/bufferViews are...

ptrgags replied:
No, we're just going to describe this inline. If possible we want only implicit tiling to need buffer views.

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+- Efficient traversal: The binary representation of tile locations in the tree level allow for easy traversal of the tileset (finding parent and child nodes).
+- Locality of reference: Adjacent indices are stored close to each other in memory and are close to each other spatially.
+
+*The following section is non-normative*
```
Put this in a subsection like `#### Morton Order Example`

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+
+*The following section is non-normative*
+
+The figure below shows the tile location decomposition of the tile `3/5/1` (Level: 3, X: 5, Y: 1). We first convert the tile location to its Morton index, which is `19`. At Level 3 of a Quadtree, we'll use 6 bits to represent the binary value of the Morton index: `010011`.
```
Use tuple notation when referring to coordinates, as template URIs can be configured by the user.

```suggestion
The figure below shows the tile location decomposition of the tile `(level, x, y) = (3, 5, 1)`. We first convert the tile location to its Morton index, which is `19`. At Level 3 of a Quadtree, we'll use 6 bits to represent the binary value of the Morton index: `010011`.
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+
+![Morton Order](figures/morton-indexing.png)
+
+### Availability Encoding
```
Oh I just realized we never specified alignment rules. Can we inherit them from the metadata specification?

sanjeetsuhag replied:
Yes, unless we have a specific reason to deviate from them.

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+        "uri": "subtrees/${level}/${x}/${y}/subtree.json"
+      },
+      "content": {
+        "mimeType": "model/gltf+json",
+        "uri": "terrain/${level}/${x}/${y}.gltf"
```
Let's not refer to glTF, stick to core 3D tiles formats.

```suggestion
        "mimeType": "application/octet-stream",
        "uri": "terrain/${level}/${x}/${y}.b3dm"
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
````diff
@@ -1,37 +1,739 @@
+          |__ model.pnts
+```
+
+OUTLINE:
+- double-headed quadtree as explicit root tileset + external implicit tilesets
+- sparse, shallow octree
+- deep, sparse quadtree
+- Full quadtree (medium depth)
````
We should remove this

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,739 @@
+* **tiling scheme** - A well-defined method for subdividing a bounding volume into a hierarchy of tiles.
+* **unavailable tile** - A region of space that does not contain a tile.
+
+## Examples
```
I feel like overall we could do better with examples if we had a more visual way of presenting at least one of them. Though as @ErixenCruz and I learned on Friday, putting together labeled examples can be time consuming (since determining morton order is necessary in some diagrams.

ptrgags reviewed (commented):
Found several small revisions, but nothing huge.

ErixenCruz commented on extensions/3DTILES_implicit_tiling/0.0.0/schema/3DTILES_implicit_tiling.schema.json:
```diff
@@ -23,19 +23,19 @@
+      "subtreeLevels": {
         "type": "integer",
-        "description": "The depth of each subtree.",
+        "description": "The number of distinct levels in each subtree. For example, a quadtree with `subtreeLevels = 2` will have subtrees with up to 5 nodes (one root and 4 children)",
```
```suggestion
        "description": "The number of distinct levels in each subtree. For example, a quadtree with `subtreeLevels = 2` will have subtrees with 5 nodes (one root and 4 children)",
```

ptrgags commented:
Note: we should be careful about subtree diagrams. subtree leaf nodes should point to the _subtree_ below them, not a node directly.

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+
+## Content
+
+**Content** is an optional 3D model associated with a single tile, as described in the [Introduction section](https://github.com/CesiumGS/3d-tiles/tree/master/specification#introduction) of the Cesium 3D Tiles 1.0 specification, with the addition of an explicit `mimeType` to reliably identify the type of content.
```
It's unfortunate that b3dm, i3dm, pnts, and cmpt don't have unique mime types yet which makes it hard to actually identify the type of content. CC https://github.com/CesiumGS/3d-tiles/issues/341

ptrgags replied:
On the other hand, if we convert our formats to glTF, they'd all have the same `mimeType` anyway... See https://github.com/CesiumGS/3d-tiles-next/issues/91#issuecomment-761122165. still good to have the mimeType though, but might be easier to enforce in 2.0.

ptrgags replied:
Resolving as there's nothing actionable until 2.0

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+
+![exact cover](figures/union-of-subtrees.jpg)
+
+A subtree has a fixed number of levels defined by the `subtreeLevels` property. This describes the number of distinct levels in the tree. The number of children per tile is also fixed due to the subdivision scheme. For quadtrees, there are `4` children per tile, while octrees have `8` children per tile. Taken together, a subtree has exactly enough tiles to store a full quadtree or full octree wit ...
```
```suggestion
A subtree has a fixed number of levels defined by the `subtreeLevels` property. This describes the number of distinct levels in the subtree. The number of children per tile is also fixed due to the subdivision scheme. For quadtrees, there are `4` children per tile, while octrees have `8` children per tile. Taken together, a subtree has exactly enough tiles to store a full quadtree or full octree with a limited number of levels. However, each tile may or may not exist or contain content, as a tileset only stores the tiles that are necessary.
```

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+
+A subtree has a fixed number of levels defined by the `subtreeLevels` property. This describes the number of distinct levels in the tree. The number of children per tile is also fixed due to the subdivision scheme. For quadtrees, there are `4` children per tile, while octrees have `8` children per tile. Taken together, a subtree has exactly enough tiles to store a full quadtree or full octree wit ...
+
+![subtree anatomy](figures/subtree-anatomy.jpg)
```
This diagram might be slightly more readable if the arrows and the outer triangle are thinner

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+**Availability** is a boolean that defines whether a tile, content, or subtree exist in a tileset. Availability serves two purposes:
+
+1. It provides an efficient method for checking what files are present
+2. Including ths information prevents extraneous HTTP requests that would result in 404 errors.
```
```suggestion
2. Including this information prevents extraneous HTTP requests that would result in 404 errors.
```

lilleyse replied:
The second bullet seems more like an example of the first bullet rather than a use case by itself

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+1. It provides an efficient method for checking what files are present
+2. Including ths information prevents extraneous HTTP requests that would result in 404 errors.
+
+Availability takes the form of a bitstream with one bit per node in consideration. A 1 indicates that a tile/content/subtree is available at this node. Meanwhile, a 0 indicates that no tile/content/subtree is available at this node.
```
Even though availability works the same way tile/content/subtree it might be confusing to describe them all together like this because it makes me think each node in the tree has subtree availability.

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+
+Availability takes the form of a bitstream with one bit per node in consideration. A 1 indicates that a tile/content/subtree is available at this node. Meanwhile, a 0 indicates that no tile/content/subtree is available at this node.
+
+For cases where all nodes are available or all nodes are unavailable, the `constant` property should be used instead of storing a full bitstream. For example, `constant: 0` indicates that all bits are 0 and no bitstream must be stored.
```
A JSON example might be helpful here. `constant: 1` might be a better example because it's a more common use case for the tree to be full rather than empty, though `constant: 0` does have its use cases too, like a tree where content only shows up deeper in the tree.

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+
+**Tile availability** is a bitstream that determines which tiles exist within a subtree. There is one bit for each subtree node. A 1 indicates that a tile is available, while a 0 indicates that a tile is unavailable.
+
+![Tile Availability](figures/tile-availability.jpg)
```
The letters on the nodes are a little small and hard to read. Maybe instead of a gradient palette use a palette of different light colors.

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+
+**Tile availability** is a bitstream that determines which tiles exist within a subtree. There is one bit for each subtree node. A 1 indicates that a tile is available, while a 0 indicates that a tile is unavailable.
+
+![Tile Availability](figures/tile-availability.jpg)
```
Just to confirm, is the tile availability array in morton order here? For binary trees is morton order the same as being ordered left-to-right?

ErixenCruz replied:
Yes I believe so.

ptrgags replied:
trivally, yes binary trees are in morton order when numbered from left-to-right. `mortonIndex = interleave(z, y, x)` but if you only have a single `x` coordinate, than `mortonIndex = interleave(x) = x`

That said, this gets back to the question of whether we should use binary trees at all in the spec. They're nice for illustration purposes since they fit better on the page, quadtrees tend to require wide diagrams. That said, in order to see the morton order, it really is non-obvious unless you use a quadtree...

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+
+In the diagram above, colored nodes indicate available tiles, while nodes with dashed outlines are unavailable tiles.
+
+If a tile is marked as available, more information about the tile, such as its content or children can be queried. If a tile is marked as unavailable, the tile must be skipped.
```
There are some rules that are included in the schema that should also be mentioned in the spec text. One example is "If a non-root tile's availability is 1 its parent tile's availability must also be 1."

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
````diff
@@ -1,37 +1,849 @@
+
+The purpose of content availability is to check if a file exists before making a network request. If content is marked as unavailable, the network request for that file must be skipped.
+
+A content availability bit must only be set if the corresponding tile availability bit is set. Otherwise, it would be possible to specify content files that are not reachable by the tiles of the tileset. The content availability bitstream can be validated by checking that the following equation holds true:
+
+```
+contentAvailability & ~tileAvailability === 0
+```
````
Personally I don't think the the validation equation needs to be included since the explanation in the previous sentences is already pretty clear.

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+
+### Child Subtree Availability
+
+**Child subtree availability** is a bitstream that determines what subtrees can be reached from this subtree. There are `N` bits for every node in the bottom-most level of the subtree, where `N` is the branching factor of the tree. A 1 means there is a child subtree available at that position in the tree. Meanwhile, a 0 means there is no subtree available.
```
"what" -> "which"?

```suggestion
**Child subtree availability** is a bitstream that determines which subtrees can be reached from this subtree. There are `N` bits for every node in the bottom-most level of the subtree, where `N` is the branching factor of the tree. A 1 means there is a child subtree available at that position in the tree. Meanwhile, a 0 means there is no subtree available.
```

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+
+**Availability** is a boolean that defines whether a tile, content, or subtree exist in a tileset. Availability serves two purposes:
+
+1. It provides an efficient method for checking what files are present
```
"what" -> "which"?

```suggestion
1. It provides an efficient method for checking which files are present
```

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+
+Availability bitstreams are packed in binary using the format described in the [Boolean Data section](https://github.com/CesiumGS/3d-tiles/blob/3d-tiles-next/specification/Metadata/0.0.0/README.md#boolean-data) of the Cesium 3D Metadata Specification. These bitstreams are subject to alignment requirements described in the [Binary Alignment Rules section](https://github.com/CesiumGS/3d-tiles/tree/ ...
+
+Each availability bitstream must be stored as a separate `bufferView`, but multiple `bufferViews` may be stored in a single `buffer`.
```
```suggestion
Each availability bitstream must be stored as a separate `bufferView`, but multiple `bufferViews` may refer to a single `buffer`.
```

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+| `boundingVolume` | a bounding volume (either a `box` or `region`) describing the root tile |
+| `refine` | Either `ADD` or `REPLACE` as in the Cesium 3D Tiles 1.0 Specification. |
+| `geometricError` | Geometric error of the root tile as described in the Cesium 3D Tiles 1.0 Specification. |
+| `maximumLevel` | Maximum level of the entire tree |
```
Is this actually the maximum level of the last available tile? Or last available content?

ptrgags replied:
tile. I updated the text.

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+| Property | Description |
+| ------ | ----------- |
+| `tilingScheme` | Either `QUADTREE` or `OCTREE`|
+| `boundingVolume` | a bounding volume (either a `box` or `region`) describing the root tile |
```
```suggestion
| `boundingVolume` | A bounding volume (either a `box` or `region`) describing the root tile |
```

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+
+| Option | Description |
+| ------ | ----------- |
+| `subtrees.uri` | template URI for a subtree JSON file. see [Subtrees](#subtrees) for more info |
```
```suggestion
| `subtrees.uri` | Template URI for a subtree JSON file. See [Subtrees](#subtrees) for more info |
```

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+| Option | Description |
+| ------ | ----------- |
+| `subtrees.uri` | template URI for a subtree JSON file. see [Subtrees](#subtrees) for more info |
+| `content.uri` | template URI for the content 3D Models |
```
Be careful not to use "3D Models" terminology too much because someday we may support other content types like rasters

```suggestion
| `content.uri` | Template URI for the content |
```

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
````diff
@@ -1,37 +1,849 @@
+| `content.uri` | template URI for the content 3D Models |
+
+Below is a full example of how the tileset JSON file looks in practice:
+
+```json
+{
+  "asset": {
+    "version": "1.0"
+  },
+  "geometricError": 10000,
+  "extensionsUsed": [
+    "3DTILES_implicit_tiling",
+  ],
+  "extensionsRequired": [
+    "3DTILES_implicit_tiling",
+  ],
+  "extensions": {
+    "3DTILES_implicit_tiling": {
+      "extras": {
+        "draftVersion": "0.0.0"
+      },
+      "boundingVolume": {
+        "region": [-1.318, 0.697, -1.319, 0.698, 0, 20]
+      },
+      "refine": "REPLACE",
+      "tilingScheme": "QUADTREE",
+      "geometricError": 5000,
+      "subtreeLevels": 7,
+      "maximumLevel": 21,
+      "subtrees": {
+        "uri": "subtrees/${level}/${x}/${y}/subtree.json"
+      },
+      "content": {
+        "mimeType": "application/octet-stream",
+        "uri": "terrain/${level}/${x}/${y}.b3dm"
+      }
+    }
+  }
+}
+```
````
It would be helpful to also see a subtree JSON alongside this

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+      "tilingScheme": "QUADTREE",
+      "geometricError": 5000,
+      "subtreeLevels": 7,
+      "maximumLevel": 21,
```
While it's totally valid for `maximumLevel` to be 20 or 21, I think the example should be 20 so that three subtrees neatly fit into the tileset JSON. With 21 there is spillover into a fourth layer of subtrees.

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+      "tilingScheme": "QUADTREE",
+      "geometricError": 5000,
+      "subtreeLevels": 2,
+      "maximumLevel": 4,
```
Same as the comment above, consider changing this to 3

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
````diff
@@ -1,37 +1,849 @@
+
+`subtrees/0/0/0/0/subtree.json`
+
+```json
+{
+	"buffers": [
+    {
+      "uri": "availability.bin",
+      "byteLength": 74
+    }
+  ],
+  "bufferViews": [
+    {
+      "buffer": 0,
+      "byteOffset": 0,
+      "byteLength": 10
+    },
+    {
+      "buffer": 0,
+      "byteOffset": 10,
+      "byteLength": 64
+    }
+  ],
+  "tileAvailability": {
+    "bufferView": 0
+  },
+  "contentAvailability": {
+    "constant": 0
+  },
+  "childSubtreeAvailability": {
+    "bufferView": 1
+  }
+}
+```
````
The formatting here is a little off

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
````diff
@@ -1,37 +1,849 @@
+        |__ 0/
+        |__ subtree.json
+        |__ availability.bin
+|__ models/
+  |__ 8/
+    |__ 0/
+      |__ 0/
+        |__ 0/
+          |__ model.pnts
+        |__ 1/
+          |__ model.pnts
+      |__ 1/
+        |__ 0/
+          |__ model.pnts
+        |__ 1/
+          |__ model.pnts
+    |__ 1/
+      |__ 0/
+        |__ 0/
+          |__ model.pnts
+        |__ 1/
+          |__ model.pnts
+      |__ 1/
+        |__ 0/
+          |__ model.pnts
+        |__ 1/
+          |__ model.pnts
+```
````
While this is valid, a more common template uri would be `"models/${level}/${z}/${y}/${x}.pnts"`. It's a little weird to have so many `model.pnts` files in practice. It also makes this diagram bigger than it needs to be.

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+| -------- | ------- | ----------- |
+| `N` | 4 or 8 | N is 4 for quadtrees, 8 for octrees |
+| `bits` | `log2(N)` | Quadtree address are a multiple of 2 bits, Octrees use a multiple of 3 bits | 
+| `mortonIndex` | `interleave(z, y, x)` or `interleave(y, x)` | The morton index is computed by interleaving bits. see below. |
```
```suggestion
| `mortonIndex` | `interleave(z, y, x)` or `interleave(y, x)` | The morton index is computed by interleaving bits. See below. |
```

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+| `N` | 4 or 8 | N is 4 for quadtrees, 8 for octrees |
+| `bits` | `log2(N)` | Quadtree address are a multiple of 2 bits, Octrees use a multiple of 3 bits | 
+| `mortonIndex` | `interleave(z, y, x)` or `interleave(y, x)` | The morton index is computed by interleaving bits. see below. |
+| `length(mortonIndex)` | `level * bits` | length of morton index in bits
```
```suggestion
| `length(mortonIndex)` | `level * bits` | Length of morton index in bits
```

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+| `index` | `(N^level - 1)/(N - 1) + mortonIndex` | Find the index of a node from `(level, mortonIndex)`
+| `level` | `ceil(log(index + 1)/log(N))` | Find the level of a node relative to the subtree |
+| `globalLevel` | `level + subtreeRoot.globalLevel` | Find the level of a node relative to the entire tileset | 
+| `startOfLevel` | `(N^level - 1)/(N - 1)` | first index at a particular level (relative to the subtree root) |
```
```suggestion
| `startOfLevel` | `(N^level - 1)/(N - 1)` | First index at a particular level (relative to the subtree root) |
```

lilleyse commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+| `lengthBits` | `N^subtreeLevels` | Length of the buffer by subtree levels |
+| `lengthBytes` | `ceil(lengthBits / 8)` | Bytes needed to store the buffer |
+| `childSubtree.globalLevel` | `subtreeRoot.globalLevel + subtreeLevels` | Level of the child subtrees relative to the tileset root |
+| `leaf.children[k].index` | `N * leaf.mortonIndex + k` | index of the `k-th` child subtree |
```
```suggestion
| `leaf.children[k].index` | `N * leaf.mortonIndex + k` | Index of the `k-th` child subtree |
```

lilleyse reviewed (commented):
I started looking at this last week and @IanLilleyT joined in since he was interested in how this relates to his work. Here's comments from both of us.

IanLilleyT commented:
Are `subdivisonScheme` and `tilingScheme` the same thing? If so, which is the right one to use?

ErixenCruz commented:
Yes they are the same. we used `tilingScheme` at first but then decided to use subdivision instead since its clearer. havent changed all the mentions of `tilingScheme` yet

IanLilleyT commented:
We should consider doing `{...}` instead of `${...}` for URI variable names.

`content/${level}/${z}/${y}/${x}.pnts` ->  `content/{level}/{z}/{y}/{x}.pnts`

This is what [CesiumJS Resource](https://cesium.com/docs/cesiumjs-ref-doc/Resource.html) expects. It's also used by terrain's `layer.json`.

[mapserver does it like this way too](https://mapserver.org/mapcache/config.html#cache)

IanLilleyT commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,845 @@
+    }
+  }
+}
+{
```
copy paste mistake?

IanLilleyT commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,845 @@
+    "extensions": {
+      "3DTILES_implicit_tiling": {
+        ...
+      }
```
Missing `}` on line below

IanLilleyT commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
````diff
@@ -1,37 +1,845 @@
+## Tileset JSON
+Like in [3D Tiles 1.0](https://github.com/CesiumGS/3d-tiles/tree/master/specification#tileset-json), one main tileset JSON file is the entry point for defining an implicit tileset. To use implicit tiling, the `3DTILES_implicit_tiling` extension must be defined in the root tile of the `tileset.json`.
+```json
+{
````
the tileset geometricError is missing

IanLilleyT replied:
This problem shows up in a few other places within the document

IanLilleyT commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
````diff
@@ -1,37 +1,845 @@
+
+The root tileset JSON might look something this:
+```json
+{
````
This tileset is missing `"root"`

IanLilleyT commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
````diff
@@ -1,37 +1,845 @@
+`tileset.json`:
+
+```json
+{
````
also missing `"root"`

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,845 @@
+
+A **Template URI** is a URI pattern used to refer to tiles by their tile coordinates.
+
+Template URIs are configured in the tileset.json. They may be any URI pattern, but must include the variables `{level}`, `{x}`, `{y}`. Template URIs for octrees must also include `${z}`. When referring to a specific tile, the tile's coordinates are substituted in for these variables.
```
the `${z}` should be `{z}`

```suggestion
Template URIs are configured in the tileset.json. They may be any URI pattern, but must include the variables `{level}`, `{x}`, `{y}`. Template URIs for octrees must also include `{z}`. When referring to a specific tile, the tile's coordinates are substituted in for these variables.
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,845 @@
+        "subdivisionScheme": "QUADTREE",
+        "content": {
+          "mimeType": "application/octet-stream",
+          "uri": "terrain/${level}/${x}/${y}.b3dm"
```
```suggestion
          "uri": "terrain/{level}/{x}/{y}.b3dm"
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,845 @@
+        "subdivisionScheme": "QUADTREE",
+        "content": {
+          "mimeType": "application/octet-stream",
+          "uri": "terrain/${level}/${x}/${y}.b3dm"
```
```suggestion
          "uri": "terrain/{level}/{x}/{y}.b3dm"
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,845 @@
+      "geometricError": 5000,
+      "subtreeLevels": 2,
+      "maximumLevel": 4,
+      "subtrees": "subtrees/${level}/${x}/${y}/subtree.json",
```
```suggestion
      "subtrees": "subtrees/{level}/{x}/{y}/subtree.json",
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,845 @@
+      "subtrees": "subtrees/${level}/${x}/${y}/subtree.json",
+      "content": {
+        "mimeType": "application/octet-stream",
+        "uri": "models/${level}/${x}/${y}.b3dm"
```
```suggestion
        "uri": "models/{level}/{x}/{y}.b3dm"
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,845 @@
+      "subtreeLevels": 3,
+      "maximumLevel": 8,
+      "subtrees": {
+        "uri": "subtrees/${level}/${z}/${y}/${x}/subtree.json"
```
```suggestion
        "uri": "subtrees/{level}/{z}/{y}/{x}/subtree.json"
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,845 @@
+      },
+      "content": {
+        "mimeType": "application/octet-stream",
+        "uri": "models/${level}/${z}/${y}/${x}/model.pnts"
```
```suggestion
        "uri": "models/{level}/{z}/{y}/{x}/model.pnts"
```

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/example/sparseOctree/tileset.json:
```diff
@@ -0,0 +1,42 @@
+      "subtreeLevels": 7,
+      "maximumLevel": 14,
+      "subtrees": {
+        "uri": "subtrees/${level}/${x}/${y}/${z}.json"
```
This goes for all the other tileset JSONs.

```suggestion
        "uri": "subtrees/{level}/{x}/{y}/{z}.json"
```

ptrgags replied:
@ErixenCruz did you generate any of these tilesets via a script? if so, could you check the script in so we can update it to fix this change?

ErixenCruz replied:
No, I did not use a script for this.

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/schema/3DTILES_implicit_tiling.schema.json:
```diff
@@ -4,38 +4,24 @@
         "minimum" : 0
       },
       "subtrees": {
         "allOf": [{"$ref": "templateUri.schema.json"}],
```
not in this file, but in `templateUri.schema.json`: Update it to use `{x}` syntax.

ptrgags reviewed (commented):
It seems like most of us want `{x}` rather than `${x}` so pointing out some places where this needs to be updated. there were many, so check for things I missed.

IanLilleyT commented:
Should
```
  "tileAvailability": {
    "constant": 0
  }
```
be explicitly forbidden?

Because if it were 0, there would be no point in having a subtree in the first place.

Another benefit of this constraint is if you have tile metadata, it's guaranteed that a tile metadata buffer always exists.

ptrgags commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,849 @@
+| `y` | Along the `+y` axis of the bounding box |
+| `y` | Along the `+z` axis of the bounding box |
+
+![Box coordinates](figures/box-coordinates.jpg)
```
so far addressed the double-labeling and curved arrows. Third image (region bounding volume) needs to be redone to be more curved. But that's lower on the priority list, the focus should be on schema/spec changes.

ptrgags commented:
@IanLilleyT @sanjeetsuhag @ErixenCruz @lilleyse 

Changelist from this week's round of implicit tiling schema/spec updates:

* subtree files are binary now! the format is similar to GLB, but a bit more slimmed down (no JSON/BIN chunk headers)
* the description of buffers/bufferViews is now included in this spec since we're getting rid of 3DTILES_binary_buffers. This description includes the 8-byte alignment requirement.
* clarified some wording about the size of `childSubtreeAvailability`
* clarification: Levels are measured starting at 0 at the root implicit tile, i.e. the tile with the `3DTILES_implicit_tiling` extension. This clarification is necessary since an implicit tileset may be part of a larger explicit tileset.
* `3DTILES_implicit_tiling.content` is optional, and when present, `subtree.contentAvailability` is required. If `content` is not specified, then `contentAvailability` is disallowed.
* remove `mimeType`, it won't be useful until 2.0, and even then it's not a complete solution. 
* Fixed some examples that still were putting the extension in the tileset rather than the root tile
* Added some details about availability that were present in the schema but not the spec.

What has NOT been changed (yet):
* the text of the spec still needs work. better diagrams and examples and a couple organization details. Full details in https://github.com/CesiumGS/3d-tiles-next/issues/119#issuecomment-746879326
* the example files need to be fixed, I need to transform the `subtree.json` files into binary files. I plan to do this early next week to avoid confusion.

Any questions/comments, please let me know.

IanLilleyT commented on extensions/3DTILES_implicit_tiling/0.0.0/README.md:
```diff
@@ -1,37 +1,879 @@
+
+**Child subtree availability** is a bitstream that determines which subtrees can be reached from the deepest level of this subtree.
+
+The child subtree availability bitstream has slightly different structure than tile or content availability. There is one bit for each node in the level of the tree immediately below the subtree. That is, if the deepest subtree is level `L`, then there is one bit for every node at level `L + 1`. Since each node at level `L` has `N` children (4 for `QUADTREE` or 8 for `OCTREE`), there are `N * (L  ...
```
* `That is, if the deepest subtree`. Should it be `deepest subtree node`?
* I assume `L` is a level local to the subtree, but it could be read as a global level.
* The equation `N * (L + 1)` should be `N ^ (L + 1)`.
* No mention of morton encoding
 
I feel like this paragraph would be more clear if defined in terms of `subtreeLevels`

lilleyse commented:
To keep the in-progress extensions in lockstep we're going to merge this PR into the [3d-tiles-next](https://github.com/CesiumGS/3d-tiles/tree/3d-tiles-next) staging branch and work from there.
